renku/core/models/provenance/activity.py

```diff
@@ -0,0 +1,417 @@
+
+        # TODO: _was_informed_by = attr.ib(kw_only=True,)
+        # TODO: influenced = attr.ib(kw_only=True)
+
+    @classmethod
+    def from_process_run(cls, process_run: ProcessRun, plan: Plan, client, order=None):
+        """Create an Activity from a ProcessRun."""
+        activity_id = Activity.generate_id(client)
+
+        association = Association(agent=process_run.association.agent, id=f"{activity_id}/association", plan=plan)
+
+        # NOTE: The same entity can have the same id during different times in its lifetime (e.g. different commit_sha,
+        # but the same content). When it gets flattened, some fields will have multiple values which will cause an error
+        # during deserialization. Make sure that no such Entity attributes exists (store those information in the
+        # Generation object).
+
+        qualified_usage = _convert_qualified_usage(process_run.qualified_usage, activity_id, client)
+
+        generated = _convert_generated(process_run.generated, activity_id, client)
+
+        invalidated = [_convert_invalidated_entity(e, activity_id, client) for e in process_run.invalidated]
+
+        return cls(
+            agents=process_run.agents,
+            association=association,
+            ended_at_time=process_run.ended_at_time,
+            generated=generated,
+            id_=activity_id,
+            invalidated=invalidated,
+            order=order,
+            # project=process_run._project,
+            qualified_usage=qualified_usage,
+            started_at_time=process_run.started_at_time,
+            annotations=process_run.annotations,
+        )
+
+    @staticmethod
+    def generate_id(client):
+        """Generate an identifier for an activity."""
+        # TODO: make id generation idempotent
+        host = get_host(client)
+        return urljoin(f"https://{host}", pathlib.posixpath.join("activities", str(uuid.uuid4())))
+
+    @classmethod
+    def from_jsonld(cls, data):
+        """Create an instance from JSON-LD."""
+        if isinstance(data, cls):
+            return data
+        if not isinstance(data, list):
+            raise ValueError(data)
+
+        return ActivitySchema(flattened=True).load(data)
+
+    def to_jsonld(self):
+        """Create JSON-LD."""
+        return ActivitySchema(flattened=True).dump(self)
+
+
+def _convert_qualified_usage(qualified_usage: List[Usage], activity_id, client) -> List[Usage]:
+    """Convert a qualified Usages."""
+    usages = []
+
+    for usage in qualified_usage:
+        commit_sha = _extract_commit_sha(entity_id=usage.entity._id)
+        entity = _convert_usage_entity(usage.entity, commit_sha, activity_id, client)
+        assert entity, f"Top entity was not found for Usage: {usage._id}, {usage.entity.path}"
+
+        id_ = pathlib.posixpath.join("usage", str(uuid.uuid4()), entity.checksum, quote(entity.path))
+        usage_id = f"{activity_id}/{id_}"
+        duplicates_found = any([u for u in usages if u._id == usage_id])
```

Comment:
Right, before I was using `usage.role` instead of the uuid and we could have duplicates. What do you think if we should use `usage.role` or uuid?

I believe we should create two usages in this case because they have different roles (input position 1 vs. 2).